examples/spark_udf/spark_udf_with_prebuilt_env.py

```diff
@@ -0,0 +1,39 @@
+"""
+This example code shows how to use `mlflow.pyfunc.spark_udf` with Databricks Connect
+outside Databricks runtime.
+"""
+
+from databricks.connect import DatabricksSession
+from sklearn import datasets
+from sklearn.neighbors import KNeighborsClassifier
+
+import mlflow
+
+spark = DatabricksSession.builder.remote(
+    host="<Databricks host>",
+    token="<Databricks token>",
+    cluster_id="<cluster id>",  # get cluster id by spark.conf.get("spark.databricks.clusterUsageTags.clusterId")
+).getOrCreate()
+
+X, y = datasets.load_iris(as_frame=True, return_X_y=True)
+model = KNeighborsClassifier()
+model.fit(X, y)
+
+with mlflow.start_run():
+    model_info = mlflow.sklearn.log_model(model, "model")
+
+model_uri = model_info.model_uri
+
+# The prebuilt model environment archive file path.
```

Comment:
how can we build it?

Reply:
I have commented:
```
# To build the model environment, run `mlflow.pyfunc.build_model_env` in Databricks runtime,
# and then download the built env archive file to local machine.
```
user needs to run `mlflow.pyfunc.build_model_env` in Databricks runtime, but this file is run in user's own machine , which is outside databricks runtime.

Reply:
How does the environment built on Databricks runtime (Ubuntu) work on my machine (Mac)?

Reply:
> How does the environment built on Databricks runtime (Ubuntu) work on my machine (Mac)?

The prebuilt env is used in UDF, and UDF runs in UDF sandbox which has consistent runtime environment where building the env.

> How can I download the built env archive?

In databricks runtime, user can save it to  UC volume (which is mounted to /Volumes/...), and in user's own machine download from it.

Reply:
where does the UDF run? in my laptop? or databricks runtime?

Reply:
The downloaded archive will be uploaded to Databricks?

Reply:
> The downloaded archive will be uploaded to Databricks?

Yes. DBconnect client uploads the environment to NFS, then it can be accessed inside UDF sandbox.

Reply:
But it is efficient, it only needs to upload once, then it will be cached in NFS until the DBconnect session dies.

Reply:
Do I repeat this process if I log a new model?

Reply:
If 2 models have the same dependencies and python version requirement, then they share the same model environment. Otherwise you need to build another model environment.

Reply:
is there any better long term solution?

Reply:
@harupy 
I have synced with Serverless team, this https://docs.google.com/document/d/1YaI4uYMyQeCUeHCZVZPW_w2_q5pgDnp9qMFBTfExAFI/edit?tab=t.0#bookmark=id.orgjig83aykk might be the long term solution, but the major blocker is **it can't set up specific python version env), the major reason from Serverless team is:

> We have a specific Python requirements that we ship as part of our product, and this should be homogeneous within that space. This will align with keeping the product simple, keeps security exposure low, clear support SLAs, etc. We don't want to ship a feature that runs an old version of Python that we cannot support.

Reply:
But considering the awkward Serverless UDF sandbox 1GB memory+disk capacity limitation, for the long term, building distributed ML feature on top of Serverless seems not the right direction.

So for the long term, I propose to use Serverless Ray to do this.

Reply:
I see, do we need to implement this in MLflow? The current approach is not convenient at all and complicates the code.

Reply:
for long term, we should have Serverless Ray and will integrate it with MLflow.

for now, we still need spark_udf with Serverless / DBconnect integration (requested by feature store team and autoML team)

for long term, after servereless Ray is ready, we can consider deprecate spark_udf code of this part.

Reply:
For improving current approach ease-of-use, we can consider in MLflow UI, adding a button for logged model to prebuild model env for specific runtime versions and upload them to UC volumn, and display the prebuilt model env UC volumn path in MLflow UI, then user can use the prebuilt env when they need it.

Reply:
and prebuilt model env also has benefit that it speeds up UDF execution, building env usually takes around 5 minutes which is long for user. But using prebuilt env decrease the time to about less than half minutes.

Reply:
> for now, we still need spark_udf with Serverless / DBconnect integration (requested by feature store team and autoML team)

Does the current approach work for them?

Reply:
I'm not opposed to this approach if this is the only available solution and need to unblock feature store team and autoML team.

Reply:
@BenWilson2 what are your thoughts?

Reply:
It is possible to support automation, e.g. we can support:

`mlflow.<flavor>.log_model(..., build_env=True)`,
and Databricks starts backend service to automatically build model env for this model + every databricks runtime version and uploads these model env archive files to specific UC volume path.

then in DBConnect side, when invoking MLflow `spark_udf`, automatically query specific UC volume path to get the matched prebuit env and use it.

And I think this should be a better solution than current approach (rebuilding model env every time, this wastes lots of time and resources)

Reply:
> Does the current approach work for feature store team and autoML team?

They might need to update their code too for the prebuilt env part. We can't support seamless migration for now.

but it should be the only way for now.

Reply:
> The speed is not a problem. 

Personally I don't think so. Usually the same model can be used many times to do inference, if we have to rebuild model env again when running inference each time, this seems not the proper approach.

Reply:
Btw, we can ask spark-connect team to optimize NFS file uploading routine, to minimize the overhead. 
currently, we need to download file from UC volume to DBconnect client, then upload to NFS from client,
but we can reduce the overhead by directly uploading file in spark driver (i.e. spark driver download file from UC and upload it to cluster NFS.

Discussion thread: https://databricks.slack.com/archives/C03RS1CEKHC/p1729151102948519

Reply:
> Personally I don't think so

I agree. The speed matters, but manual steps are a bigger problem.

Reply:
Can we use databricks SDK to simplify the process?

Reply:
We can use databricks SDK to download file from UC volume. But we still need a Databricks service for this.

Reply:
I agree that the steps required here are almost a barrier to adopting the use here. 

- The return of `build_model_env` is a path to the tar gzip file. (seems reasonable to me)
- Navigating to this location in a Workspace, purely to do a UI-based download of this file is going to be very annoying. 
- Setting up a local location that this file will live in involves manually relocating from the `downloads` folder (default save location, which will quickly get filled up with many copies for prolonged use) to keep things clean (I doubt many users will do this; they will just fill their downloads folder up with loads of these files until they run out of disk space and have to manually delete old ones). 
- unpack the archive and activate it for the DBConnect session? 

It seems to me that we could file a follow-up PR that exposes a public API that automates this to a certain degree. 

Is there anything that prevents the UJ of:

Step 1: existing API functionality, we have a return path to a UC volume location for this model. 
Step 2 (executed locally on user's machine): new API that will fetch the archive using the WorkspaceClient and writing to a specific user-defined local location, returning the qualified path that can be used in `mlflow.pyfunc.spark_udf(<spark session>, model_uri, prebuild_env_path=<return of new API>)
Step 3: run inference as normal


I feel that there is an opportunity to simplify this quite a bit and prevent users from using the UI to download the archive or getting annoyed by having to write file transfer logic themselves. 

Can we do this as a followup?